src/ir_pattern.py

```diff
@@ -0,0 +1,452 @@
+import abc
+from collections import defaultdict
+from dataclasses import dataclass, field, replace
+from itertools import permutations
+from typing import ClassVar, Dict, List, Optional, TypeVar
+
+from .error import static_assert_unreachable
+from .flow_graph import (
+    ArchFlowGraph,
+    BaseNode,
+    FlowGraph,
+    InstrRef,
+    Instruction,
+    LocationRefSetDict,
+    RefSet,
+    Reference,
+    TerminalNode,
+    build_flowgraph,
+)
+from .parse_file import AsmData, Function
+from .parse_instruction import (
+    Argument,
+    AsmAddressMode,
+    AsmGlobalSymbol,
+    AsmInstruction,
+    AsmLiteral,
+    BinOp,
+    Instruction,
+    InstructionMeta,
+    JumpTarget,
+    Location,
+    RegFormatter,
+    Register,
+    StackLocation,
+    parse_instruction,
+)
+
+
+@dataclass(eq=False, frozen=True)
+class IrPattern(abc.ABC):
+    """
+    Template for defining "IR" patterns that can match against input asm.
+    The matching process uses the FlowGraph and register analysis to compute
+    inter-instruction dependencies, so these patterns can match even when
+    they have been interleaved/reordered by the compiler in the input asm.
+
+    IrPattern subclasses *must* define `parts` and `replacement`, and can
+    optionally implement `check()`.
+
+    For now, the pattern cannot contain any branches, and the replacement
+    must be a single instruction (though, it can be fictive).
+    """
+
+    parts: ClassVar[List[str]]
+    replacement: ClassVar[str]
+
+    def check(self, m: "IrMatch") -> bool:
+        """Override to perform additional checks/calculations before replacement."""
+        return True
+
+    def compile(self, arch: ArchFlowGraph) -> "CompiledIrPattern":
+        missing_meta = InstructionMeta.missing()
+        regf = RegFormatter()
+        replacement_instr = parse_instruction(
+            self.replacement, missing_meta, arch, regf
+        )
+
+        name = f"__pattern_{self.__class__.__name__}"
+        func = Function(name=name)
+        # Add a fictive nop instruction for each input to the replacement_instr
+        # This acts as a placeholder Reference to represent where the input was set
+        assert len(replacement_instr.inputs) == len(
+            set(replacement_instr.inputs)
+        ), "pattern inputs must be unique"
+        for inp in replacement_instr.inputs:
+            func.new_instruction(
+                Instruction(
+                    "in.fictive",
+                    [],
+                    meta=missing_meta,
+                    inputs=[],
+                    clobbers=[],
+                    outputs=[inp],
+                    reads_memory=False,
+                    writes_memory=False,
+                )
+            )
```

Comment:
Do we really need in.fictive instructions, or can we use the inputs in place of/addition to `all_registers` when creating the pattern use/def graph?

My feeling is that it should be fine to replace all in.fictive instructions by a single one, since all dependency maps are keyed by location, and then replace the single remaining instruction by nothing since we know use/def-analyze even the entry state.

Reply:
The dependency maps are keyed by location, but the `TryIrMatch.ref_map` match state is not. 

I think of `ref_map` as a many-to-one relationship between pattern instructions & matched instructions, which is also why the direction of the map matters.

Reply:
I don't quite see how that matters here -- surely the in.fictive instructions are not in ref_map, since inputs to the pattern are allowed to be phis? (I should probably take another look at the code for improved understanding of how the in.fictive's are used)

Reply:
(Putting aside [this issue](https://github.com/matt-kempster/mips_to_c/pull/214#discussion_r855645733) where RefSets are used ambiguously both for phis & overlapping assignments)

The in.ficitve instructions are in ref_map. For an example like the following:

```asm
# replacement: example.fictive $z, $a, $b
fadds $x, $a, $b
fadds $y, $a, $b
fadds $z, $x, $y
```

During matching, we need to check that both uses of `$a` come from the same Reference, and same for `$b`. But, `$a` can come from a different Reference than `$b`. 
The `in.fictive $a` (which has no inputs and outputs only `$a`) acts as placeholder in `ref_map` so that we can check that all uses of `$a` come from the same place. 

You're right, however, that the matching needs to be stricter when phis are involved. There probably needs to be a bigger change made to handle inputs differently than internal registers (so that only inputs can match phis).